building/build-debs/homeworld-etcd-metrics-exporter/src/etcd-metrics-exporter.go

```diff
@@ -0,0 +1,212 @@
+/*
+Copyright 2018 Cel Skeggs (modifications)
+Copyright 2017 The Kubernetes Authors (original)
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	"log"
+	"os"
+	"strings"
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"io"
+)
+
+// Initialize the prometheus instrumentation and client related flags.
+var (
+	etcdScrapeBase string
+
+	httpClient *http.Client
+)
+
+// Initialize prometheus metrics to be exported.
+var (
+	// Register all custom metrics with a dedicated registry to keep them separate.
+	customMetricRegistry = prometheus.NewRegistry()
+
+	// Custom etcd version metric since etcd 3.2- does not export one.
+	// This will be replaced by https://github.com/coreos/etcd/pull/8960 in etcd 3.3.
+	etcdVersion = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "etcd",
+			Name:      "version_info",
+			Help:      "Etcd server's binary version",
+		},
+		[]string{"binary_version"})
+)
+
+// monitorGatherer is a custom metric gatherer for prometheus that exports custom metrics
+// defined by this monitor as well as rewritten etcd metrics.
+type monitorGatherer struct {
+}
+
+func (m *monitorGatherer) Gather() ([]*dto.MetricFamily, error) {
+	etcdMetrics, err := scrapeMetrics()
+	if err != nil {
+		return nil, err
+	}
+	custom, err := customMetricRegistry.Gather()
+	if err != nil {
+		return nil, err
+	}
+	result := make([]*dto.MetricFamily, 0, len(etcdMetrics)+len(custom))
+	for _, mf := range etcdMetrics {
+		result = append(result, mf)
+	}
+	result = append(result, custom...)
+	return result, nil
+}
+
+// Struct for unmarshalling the json response from etcd's /version endpoint.
+type EtcdVersion struct {
+	BinaryVersion  string `json:"etcdserver"`
+	ClusterVersion string `json:"etcdcluster"`
+}
+
+func requestToEtcd(endpoint string) (io.ReadCloser, error) {
+	if !strings.HasPrefix(endpoint, "/") {
+		return nil, fmt.Errorf("no '/' prefix on endpoint")
+	}
```

Comment:
I tend to prefer being explicit with strings like "/metrics" and "/version" instead of "metrics" and "version", because it makes it clear that they're paths, and not just identifiers.

I could make it so that the base URL also has a trailing slash, and then remove the adjacent slashes?